WIP: Add support for comparing more than two benchmarks.

robinheghan opened the pull request:
Every now and again, I've had the need to compare more than two benchmarks. For instance, what is faster of the Core list implementation, my new list implementation, and this altered version of my new list implementation.

This PR adds support for comparing as many benchmarks as the user sees fit.

This PR might need some more work, so I hope the all wise Brian can give me some pointers.

* I'm not sure of the UI. It takes a lot of (horizontal) space.
* Instead of a `compare` fn taking a list, maybe it would be better to have `compare2`, `compare3`, `compare4`... This way, a comparison has to be made with at least two benchmarks.
* The code can probably be cleaned up a little.

I'm thinking of adding `elm-plot` diagrams to this, but it's nice to agree on the base implementation first.

robinheghan committed:
Add support for comparing more than two benchmarks.

BrianHicks commented on src/Benchmark.elm:
```diff
@@ -212,10 +211,19 @@ benchmark function**. For example, use only `benchmark2` instead of mixing
+            Internal.Group name []
+
+        [ benchmark ] ->
+            benchmark
```
These two cases are why `compare` explicitly takes two benchmarks at the moment. This increases the possibility that the user will use the API incorrectly and get bad results. It sounds like your use case needs `compare3`, not `compareListOfPotentitallyInfiniteOrZeroSize`. It might be better to implement `compare3` instead.

BrianHicks reviewed (changes requested):
Thanks for the PR!

1. Can you attach a screenshot of the UI? How are you comparing things?
2. This doesn't compile. Check out those Travis results. Looks mostly like tests and examples don't work because the API changed. I'd like to see how the ergonomics of the API change because of this, so the examples are especially interesting.

robinheghan committed:
Add comparison functions 2 through 8.

robinheghan committed:
Fix examples to use new comparison fns.

robinheghan committed:
Fixed tests.

robinheghan commented:
So I implemented compare3 -> compare8, and renamed compare to compare2 to match the benchmarkN api.

Personally, I can imagine myself using up to compare4, but who knows?

Will add an image with how it looks now.

robinheghan commented:
![elm-bench](https://user-images.githubusercontent.com/854889/27149874-5c10a7ee-5145-11e7-8a6c-4e2c395e1362.png)

BrianHicks commented:
I'm pretty sure we should enter into a separate mode when comparing more than two benchmarks… order the results by speed or something. I'm hesitant to merge this until we can present the information better.

For now, why don't we leave off 5 through 8? If you can't imagine yourself using them, why should we expand the surface area of the API? If you have a concrete case for `benchmark4` we can keep it, but if it just feels like something to have let's drop that too.

robinheghan commented:
Added compare2 -> compare8 to match the `benchmarkN` fns. I have definite need for `compare4`, but I'll drop the others.

I think it would be better to have the (std. deviation) numbers on their own row, below the relevant numbers. I could also try ordering the benchmarks by ops/second.

BrianHicks commented:
As of right now, there are merge conflicts on this branch. I don't want to leave this open and in limbo, so here's what it would take to get this merged:

1. Rebase or merge `master` into this branch and fix the conflicts (there were a lot introduced by #18 and the pre-work I did to accept that.)
2. Make sure everything compiles and the tests pass
3. Let's have a chat about the API design here. I think that's probably what you're most interested in; would you like to schedule a hangout to talk about it? This feels like it may benefit from a high-bandwidth communication channel.

If you're not interested right now, that's fine too. If that's the case, you can close the PR yourself. If I don't hear back in a week or so I'll just transfer the things you want to an issue so I or someone else can pick it up another time.

Thanks again for thinking about these things with me at Elm Europe and beyond. ❤️

robinheghan commented:
I've been meaning to revisit this, i've just been busy lately. I'll try to visit this this weekend. In any case, a hangout would be great. What's your timezone? Does sunday work?

> Den 13. jul. 2017 kl. 16.09 skrev Brian Hicks <notifications@github.com>:
> 
> As of right now, there are merge conflicts on this branch. I don't want to leave this open and in limbo, so here's what it would take to get this merged:
> 
> Rebase or merge master into this branch and fix the conflicts (there were a lot introduced by #18 and the pre-work I did to accept that.)
> Make sure everything compiles and the tests pass
> Let's have a chat about the API design here. I think that's probably what you're most interested in; would you like to schedule a hangout to talk about it? This feels like it may benefit from a high-bandwidth communication channel.
> If you're not interested right now, that's fine too. If that's the case, you can close the PR yourself. If I don't hear back in a week or so I'll just transfer the things you want to an issue so I or someone else can pick it up another time.
> 
> Thanks again for thinking about these things with me at Elm Europe and beyond. ❤️
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

BrianHicks commented:
Following up on scheduling in the Elm Slack.

robinheghan commented:
Closing this.

The use case I want supported is mentioned in #21
